File: server/src/main/java/com/example/amande/domain/models/paint/Paint.java
package com.example.amande.domain.models.paint;

import com.example.amande.domain.models.paint.hue.CIEDE2000;
import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.Value;
import lombok.experimental.Accessors;

import static lombok.AccessLevel.PRIVATE;

@AllArgsConstructor(access = PRIVATE)
@Accessors(fluent = true)
@Value
public class Paint {

  PaintID id;
  PaintName name;
  PaintColorCode colorCode;
  PaintColorProximity colorProximity;

  public static Paint reconstruct(
    @NonNull PaintID id,
    @NonNull PaintName name,
    @NonNull PaintColorCode colorCode
  ) {
    return new Paint(
      id,
      name,
      colorCode,
      new PaintColorProximity(0d)
    );
  }

  public Paint calculateColorProximity(PaintColorCode colorCode) {
    return new Paint(
      this.id,
      this.name,
      this.colorCode,
      new PaintColorProximity(CIEDE2000.calculation(this.colorCode, colorCode))
    );
  }
}

File: client/src/app/lib/models/paint.ts
export class PaintID {
  constructor(public readonly value: string) {}
}

export class PaintName {
  constructor(public readonly value: string) {}
}

export class ColorCode {
  constructor(public readonly value: string) {}
}

export class Paint {
  constructor(
    public id: PaintID,
    public paintName: PaintName,
    public colorCode: ColorCode,
  ) {}

  changePaintName(paintName: PaintName): Paint {
    return new Paint(
      this.id,
      paintName,
      this.colorCode,
    )
  }
  changeColorCode(colorCode: ColorCode): Paint {
    return new Paint(
      this.id,
      this.paintName,
      colorCode,
    )
  }

}
